src/caller.jl

```diff
@@ -0,0 +1,189 @@
+
+function prepare_argument(x::String)
+    return Any[ mapping_types_reversed[:STRING_CMD], libSingular.copy_string_to_void(x) ], nothing
+end
+
+function prepare_argument(x::PolyRing)
+    new_ptr = libSingular.get_ring_ref(x.ptr)
+    return Any[ mapping_types_reversed[:RING_CMD], new_ptr ], x
+end
+
+function prepare_argument(x::spoly)
+    rng = parent(x)
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:POLY_CMD], libSingular.copy_polyptr_to_void(x.ptr, rng_ptr) ], rng
+end
+
+function prepare_argument(x::svector)
+    rng = parent(x).base_ring
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:VECTOR_CMD], libSingular.copy_polyptr_to_void(x.ptr, rng_ptr) ], rng
+end
+
+function prepare_argument(x::sideal)
+    rng = parent(x).base_ring
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:IDEAL_CMD], libSingular.copy_idealptr_to_void(x.ptr, rng_ptr)], rng
+end
+
+function prepare_argument(x::smodule)
+    rng = parent(x).base_ring
+    rng_ptr = rng.ptr
+    return Any[ mapping_types_reversed[:MODUL_CMD], libSingular.copy_idealptr_to_void(x.ptr, rng_ptr)], rng
+end
+
+function prepare_argument(x::sresolution)
+    rng = base_ring(x)
+    res = Any[ mapping_types_reversed[:RESOLUTION_CMD], libSingular.create_syStrategy_data(x.ptr, rng.ptr) ]
+    return res, rng
+end
+
+function prepare_argument(x::Any)
+    if x.ptr isa libSingular.number
+        ptr = x.ptr
+        rng = parent(x)
+        new_ptr = libSingular.n_Copy(ptr, rng.ptr)
+        return Any[ mapping_types_reversed[:NUMBER_CMD], new_ptr.cpp_object ], nothing
+    elseif x.ptr isa libSingular.ip_smatrix
+        rng = parent(x)
+        return Any[ mapping_types_reversed[:MATRIX_CMD], libSingular.mpCopy(x.ptr, rng.ptr).cpp_object ], rng
+    elseif x.ptr isa libSingular.__mpz_struct
+        return Any[ mapping_types_reversed[:BIGINT_CMD], x.ptr.cpp_object ], nothing
+    elseif x.ptr isa libSingular.sip_smap
+        return Any[ mapping_types_reversed[:MAP_CMD], x.ptr.cpp_object ], nothing
+    elseif x.ptr isa libSingular.bigintmat
+        return Any[ mapping_types_reversed[:BIGINTMAT_CMD], x.ptr.cpp_object ], nothing
+    end
+
+end
+
+function low_level_caller_rng(lib::String, name::String, ring, args...)
+    libSingular.load_library(lib)
+    arguments = [prepare_argument(i) for i in args]
+    arguments = Any[ i for (i, j) in arguments ]
+    return_value = libSingular.call_singular_library_procedure(name, ring.ptr, arguments)
+    return convert_return_list(return_value, ring)
+end
+
+function low_level_caller(lib::String, name::String, args...)
+    libSingular.load_library(lib)
+    arguments = [prepare_argument(i) for i in args]
+    rng = nothing
+    for (i, j) in arguments
+        if j != nothing
+            rng = j
+        end
+    end
+    arguments = Any[ i for (i, j) in arguments ]
+    return_values = nothing
+    if rng == nothing
+        return_value = libSingular.call_singular_library_procedure(name, arguments)
```

Comment:
How about this instead (assuming it works as intended, which it may not, I don't know anything about cxxwrapper):
```suggestion
        return_value = libSingular.call_singular_library_procedure(name, C_NULL, arguments)
```

Then you can get rid of `call_singular_library_procedure_wo_ring` (I find the result easier to grok, but of course that's highly subjective, so i won't complain if you leave it as is ;-)

In fact, I'd then just do this:
```julia
   rng = (rng == nothing) ? C_NULL : rng.ptr
   result = libSingular.call_singular_library_procedure(name, rng, arguments)
   return convert_return_list(result, rng)
end
```

Reply:
Unfortunately, this won't work as easy as you think here.

Because `call_singular_library_procedure` is a CxxWrapped function, the argument needs to be of the right type. There is no way we could either give a `ring` or a `C_NULL`, which is a `Ptr{Nothing}`.

We would have the following possibilities:
1. Pass the raw ring pointer to `call_singular_library_procedure` instead of the ring object. This might cause trouble if the CxxWrap's object structure ever changes (although quite some code relies on this anyway).
2. We overload this function to either take `void*` or `ring`.

I implemented the second for now.